python/python/bystro/prs/preprocess_for_prs.py

```diff
@@ -125,38 +164,60 @@ def _preprocess_genetic_maps(map_directory_path: str) -> dict[int, list[int]]:
     return bin_mappings
 
 
-def filter_scores_by_p_value(scores: pd.DataFrame, p_value_threshold: float) -> pd.DataFrame:
-    """Filter to keep rows with P-values less than the specified threshold for C+T method."""
-    return scores[scores["P"] < p_value_threshold]
+def read_feather_in_chunks(file_path, columns=None, chunk_size=1000):
+    """Read a Feather file in chunks as pandas Dataframes."""
+    table = feather.read_table(file_path, columns=columns)
+    for i in range(0, table.num_rows, chunk_size):
+        chunk = table.slice(i, chunk_size).to_pandas()
+        chunk.index = range(i, i + len(chunk))
+        yield chunk
+
+
+def get_p_value_thresholded_indices(df, p_value_threshold):
+    """Return indices of rows with P-values less than the specified threshold."""
+    return df.index[df["P"] < p_value_threshold].tolist()
 
 
 def generate_thresholded_overlap_scores_dosage(
     gwas_scores_path: str, dosage_matrix_path: str
-) -> tuple[pd.DataFrame, pd.DataFrame]:
-    """Compare and restrict to overlapping loci between dosage matrix and scores."""
+) -> pd.DataFrame:
+    """Compare and restrict to overlapping loci between dosage matrix and thresholded scores."""
     scores = _preprocess_scores(gwas_scores_path)
-    dosage_feather = _load_preprocessed_dosage_matrix(dosage_matrix_path)
+    dosage_loci_nomiss = _extract_nomiss_dosage_loci(dosage_matrix_path)
+
     # TODO: Add customizable p value threshold and option for multiple thresholds
-    thresholded_scores = filter_scores_by_p_value(scores, 0.05)
+    p_value_threshold = 0.05
+    thresholded_indices_set = set()
+    for chunk in read_feather_in_chunks(gwas_scores_path, columns=["P"], chunk_size=1000):
+        thresholded_indices = get_p_value_thresholded_indices(chunk, p_value_threshold)
+        thresholded_indices_set.update(thresholded_indices)
+    scores_filtered_indices = sorted(list(thresholded_indices_set))
+    thresholded_scores = scores.iloc[scores_filtered_indices]
+    thresholded_scores = thresholded_scores.set_index("SNPID")
+
+    if len(thresholded_scores) == 0:
+        raise ValueError("No thresholded scores available; cannot proceed with PRS calculation.")
+
     set_A = set(thresholded_scores.index)
-    set_B = set(dosage_feather["locus"])
+    set_B = set(dosage_loci_nomiss["locus"])
+
     overlap_snps = set_A.intersection(set_B)
```

Comment:
Instead of doing this, add an optional argument to `_extract_nomiss_dosage_loci`: `desired_loci`. Pass in the set of loci, which are `set(threshold_scores.index)`, and make one of the filter conditions the selection of just those loci. You will further reduce memory usage.

Reply:
This would be good, but not important for now.